Resonace Covariance Module

icmeyer opened the pull request:
This PR incorporates into the API the ability to parse resonance covariance data from ENDF File 32. Its intended functionality can be seen in the new example notebook: openmc/examples/jupyter/nuclear-data-resonance-covariance.ipynb
Namely:
- Parse File 32 data
- Allow for selection of a subset of the parameters and corresponding covariance data
- Sampling of resonance parameters (currently uses numpy.random.multivariate_normal)
- Reconstruction of the cross section using sampled resonance parameters

My first PR so comments on style encouraged!

paulromano commented:
Thanks for your first PR @icmeyer! I should be able to review it tomorrow morning.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                    sample_params = pd.DataFrame.from_records(records, columns=columns)
+                    samples.append(sample_params)
+    
+        self.samples = samples
```
The docstring says that the function returns samples. Here it actually sets it as an attribute. I would have the function `return samples` rather than setting an attribute (which is not documented). You'll need to change the function signature for `reconstruct` to accept a dataframe of sampled parameters rather than an index into the list.

paulromano replied:
@icmeyer and I just discussed having this function return a list of `ResonanceRange` objects, which can then be used to call `reconstruct()` directly.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+
+class MultiLevelBreitWignerCovariance(ResonanceCovarianceRange):
+    """Multi-level Breit-Wigner resolved resonance formalism covariance data.
+    Parameters
```
Need a blank line above this

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+        self.cov_subset = cov_subset
+
+    def sample_resonance_parameters(self, n_samples, use_subset=False):
+        """Return a IncidentNeutron object with n_samples of xs
```
This method does not return an IncidentNeutron object

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+        Returns
+        -------
+        samples : list of openmc.data.ResonanceCovarianceRange objects 
+            List of samples size [n_samples]
```
```
List of samples of size `n_samples`
```

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                        records.append([energy[j], l_value[j], spin[j], gt[j], gn[j],
+                                        gg[j], gf[j], gx[j]])
+                    columns = ['energy', 'L', 'J', 'totalWidth', 'neutronWidth',
+                           'captureWidth', 'fissionWidth', 'competitiveWidth']
```
Align with opening delimiter (see full explanation [here](https://www.python.org/dev/peps/pep-0008/?#indentation))

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                gx = pd.DataFrame.as_matrix(parameters['competitiveWidth'])
+                mean = mean_array.flatten()
+                for i in range(n_samples):
+                    sample = np.random.multivariate_normal(mean,cov)
```
space after comma: `mean, cov`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                    samples.append(sample_params)
+    
+            elif mpar == 4:
+                param_list = ['energy','neutronWidth','captureWidth','fissionWidth']
```
spaces after commas

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                    samples.append(sample_params)
+    
+            elif mpar == 5:
+                param_list = ['energy','neutronWidth','captureWidth',
```
spaces after commas

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+        # Handling RM Sampling
+        if formalism == 'rm':
+            if mpar == 3:
+                param_list = ['energy','neutronWidth','captureWidth']
```
spaces after commas

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                        records.append([energy[j], l_value[j], spin[j], gn[j],
+                                        gg[j], gfa[j], gfb[j]])
+                    columns = ['energy', 'L', 'J', 'neutronWidth',
+                           'captureWidth', 'fissionWidthA','fissionWidthB']
```
align with opening delimiter

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+        ----------
+        energies : float or Iterable of float
+            Energies at which the cross section should be evaluated
+        resonances : openmc.data.Resonance object
```
`openmc.data.ResonanceRange`?

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+        sample_parameters = self.samples[sampleN]
+        xs_array = resonances.reconstruct(energies, use_sample = True,
+                                          sample_parameters = sample_parameters)
+        return xs_array
```
`return resonances.reconstruct(...)`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+        items : list
+            Items from the CONT record at the start of the resonance range
+            subsection
+        resonances : openmc.data.IncidentNeutron.Resonance object
```
I think this should be the dataframe of parameters from a `ResonanceRange` object?

paulromano replied:
Also, needs a description below

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+
+                #Build the upper-triangular covariance matrix
+                cov_dim = mpar*num_res
+                cov = np.zeros([cov_dim,cov_dim])
```
`np.zeros((cov_dim, cov_dim))`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+        items = get_cont_record(file_obj)
+        target_spin = items[0]
+        LCOMP = items[3] # Flag for compatibility 0,1,2 - 2 is compact form
+        NLS = items[4]  # number of l-values
```
use lowercase variable names

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+            gf = values[5::12]
+            par_unc = []
+            for i in range(num_res):
+                res_unc = values[i*12+6:i*12+12]
```
`res_unc = values[12*i+6 : 12*i+12]`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+            #covariance matrix)
+            nparams,params = parameters.shape
+            covsize = cov.shape[0]
+            mpar = int(covsize/nparams)
```
`covsize // nparams` should suffice

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+
+            #Determine mpar (number of parameters for each resonance in
+            #covariance matrix)
+            nparams,params = parameters.shape
```
space after comma

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                # DAJ/DGT always zero, DGF sometimes none zero [1,2,5]
+                res_unc_nonzero = []
+                for j in range(6):
+                    if j in [1,2,5] and res_unc[j] != 0.0 :
```
`(1, 2, 5)`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                for j in range(6):
+                    if j in [1,2,5] and res_unc[j] != 0.0 :
+                        res_unc_nonzero.append(res_unc[j])
+                    elif j in [0,3,4]:
```
`(0, 3, 4)`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+            return mlbw
+
+        elif LCOMP == 0 :
+            cov = np.zeros([4,4])
```
(4, 4)

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                items, values = get_list_record(file_obj)
+                num_res = items[5]
+                for j in range(num_res):
+                    one_res = values[18*j:18*(j+1)]
```
use spaces around `:`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                    gn = res_values[3]
+                    gg = res_values[4]
+                    gf = res_values[5]
+                    records.append([energy, spin, gt, gn, gg, gf])
```
Looks like you could just: `records.append(list(res_values))`. No need to assign energy, spin, etc.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+
+                    cov_index += 4
+                    if j < num_res-1: #Pad matrix for additional values
+                        cov = np.pad(cov,((0,4),(0,4)),'constant',
```
spaces after commas

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+            #covariance matrix)
+            nparams,params = parameters.shape
+            covsize = cov.shape[0]
+            mpar = int(covsize/nparams)
```
again, `covsize // nparams`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+            mlbw.parameters = parameters
+            mlbw.covariance = cov
+            mlbw.mpar  = mpar
+            mlbw.lcomp = LCOMP
```
This block of code for creating the instance is the same on each branch of the `if`, so it can just be outside the `if` at the end. I would also change the signature of `__init__` to `__init__(self, energy_min, energy_max, parameters, covariance, mpar, lcomp)`.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+    """
+
+    def __init__(self, energy_min, energy_max):
+        super().__init__(energy_min,energy_max)
```
space after comma

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+    def __init__(self, energy_min, energy_max):
+        super().__init__(energy_min,energy_max)
+        self.formalism = 'slbw'
+
```
two blank lines before class

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+            rmc.parameters = parameters
+            rmc.covariance = cov
+            rmc.mpar  = mpar
+            rmc.lcomp = LCOMP
```
Same comment as with MLBW case: this is the same on each branch of the `if`. `__init__` method should just take all parameters and handle setting the attributes on the instance.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+            return rmc
+
+_FORMALISMS = {
+               0: ResonanceCovarianceRange,
```
Use a hanging indent here (described [here](https://www.python.org/dev/peps/pep-0008/?#indentation))

paulromano commented on tests/unit_tests/test_data_neutron.py:
```diff
@@ -39,7 +39,7 @@ def sm150():
     """Gd154 ENDF data (contains Reich Moore resonance range)"""
     filename = os.path.join(_ENDF_DATA, 'neutrons', 'n-064_Gd_154.endf')
-    return openmc.data.IncidentNeutron.from_endf(filename)
+    return openmc.data.IncidentNeutron.from_endf(filename, get_covariance = True)
```
No space around `=` when used for keyword argument

paulromano commented on tests/unit_tests/test_data_neutron.py:
```diff
@@ -96,6 +96,12 @@ def am244():
     endf_file = os.path.join(_ENDF_DATA, 'neutrons', 'n-095_Am_244.endf')
     return openmc.data.IncidentNeutron.from_njoy(endf_file)
 
+@pytest.fixture(scope='module')
```
Add extra blank line above

icmeyer commented on docs/source/conf.py:
```diff
@@ -54,7 +54,7 @@
               'sphinx.ext.viewcode',
               'sphinx.ext.imgconverter',
-              'sphinx_numfig',
+              'sphinx.ext.numfig',
```
I use conda to manage my python packages. I installed sphinx and was getting the error that `sphinx_numfig` wasn't available and I wasn't able to find it using conda. Was able to make the docs myself using the above line

paulromano commented on openmc/data/neutron.py:
```diff
@@ -289,6 +296,11 @@ def resonances(self, resonances):
+    @resonance_covariance.setter
+    def resonance_covariance(self, resonance_covariance):
+        cv.check_type('resonances', resonances, res.ResonanceCovariance)
+        self._resonacne_covariance = resonance_covariance
```
Typo "resonacne"

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,841 @@
+                gx = pd.DataFrame.as_matrix(parameters['competitiveWidth'])
+                mean = mean_array.flatten()
+                for i in range(n_samples):
+                    sample = np.random.multivariate_normal(mean,cov)
```
One more thing to consider -- it looks like the `multivariate_normal` function also takes an optional `size` argument that would allow you to generate multiple samples with a single function call. This should help improve performance when many samples are needed.

icmeyer commented:
Addressed the comments by @paulromano. Added in warning on the sampling method that it can produce nonphysical negative parameters which may cause reconstruction to fail. This is because the covariance matrix assumes a multivariate normal distribution from which negative numbers may be drawn. Further discussion and possible solutions to this issue can be found in a Rochman paper [here](http://dx.doi.org/10.1016/j.anucene.2011.01.026).

paulromano commented on openmc/data/endf.py:
```diff
@@ -71,6 +71,24 @@ def float_endf(s):
 
 
+def int_endf(s):
+    """Conver string to int. Used for INTG records where blank entries
```
Typo: convert

paulromano commented on openmc/data/endf.py:
```diff
@@ -71,6 +71,24 @@ def float_endf(s):
     return float(_ENDF_FLOAT_RE.sub(r'\1e\2', s))
 
 
+def int_endf(s):
```
Change to `_int_endf` since it is unlikely a user would ever need this

paulromano commented on openmc/data/endf.py:
```diff
@@ -249,6 +267,50 @@ def get_tab2_record(file_obj):
 
     return params, Tabulated2D(breakpoints, interpolation)
 
+def get_intg_record(file_obj):
```
Add another blank line before this function (PEP8: 2 lines between functions/classes at top level)

paulromano commented on openmc/data/endf.py:
```diff
@@ -249,6 +267,50 @@ def get_tab2_record(file_obj):
+    corr = corr + corr.T - np.diag(corr.diagonal())
+    return corr
+
+
```
delete one blank line here

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,812 @@
+        self.cov_subset = cov_subset
+
+    def sample_resonance_parameters(self, n_samples, resonances, use_subset=False):
+        """Return a list  size 'n_samples' of openmc.data.ResonanceRange objects.
```
I would say something more like "Sample resonance parameters based on the covariances." The type of the object returned (and its length) are described in the "Returns" section so probably not necessary to mention them here too.

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+    "filename, headers = urllib.request.urlretrieve(url, 'gd157.endf')\n",
+    "\n",
+    "# Load into memory\n",
+    "gd157_endf = openmc.data.IncidentNeutron.from_endf(filename, covariance = True)\n",
```
`covariance=True` (no spaces around = when used for keyword argument)

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+   ],
+   "source": [
+    "first_five = gd157_endf.resonance_covariance.ranges[0].parameters[:5]\n",
+    "print(first_five)"
```
Here you don't really need to assign to a variable and then print it. Just running:
```Python
gd157_endf.resonance_covariance.ranges[0].parameters[:5]
```
would be sufficient because Jupyter notebook will show the result of the expression

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The newly created object will contain multiple resonance regions within 'gd157_endf.res_covariance.ranges'. We can access the full covariance matrix from File 32 for a given range by:"
```
res_covariance -> resonance_covariance

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+   ],
+   "source": [
+    "plt.imshow(covariance)\n",
+    "plt.colorbar()"
```
You might want to use a [diverging colormap](https://matplotlib.org/examples/color/colormaps_reference.html) with the min/max values set to be negatives of each other (e.g., -0.008 and 0.008) so that it is obvious where zero is.

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+   ],
+   "source": [
+    "lower_bound = 2; #inclusive\n",
+    "upper_bound = 2; #inclusive\n",
```
Remove semicolons on this line and one before

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+    "upper_bound = 2; #inclusive\n",
+    "gd157_endf.resonance_covariance.ranges[0].res_subset('J',[lower_bound,upper_bound])\n",
+    "subset_first_five = gd157_endf.resonance_covariance.ranges[0].parameters_subset[:5]\n",
+    "print(subset_first_five)"
```
As before, no need for variable assignment and explicit `print` -- the last expression in a cell will get printed by default

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,812 @@
+                    res_range.parameters = sample_params
+                    samples.append(res_range)
+    
+        self.samples = samples
```
As we discussed before, the `samples` object should be returned directly rather than being set as an attribute of the current instance.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,812 @@
+        cov_subset[tri_indices] = cov_subset_vals
+    
+        self.parameters_subset = parameters_subset
+        self.cov_subset = cov_subset
```
As with the sampling method, this method should return an object rather than setting attributes of the current instance. I think what would make the most sense is to return a new instance of `ResonanceCovarianceRange` with the specified subset of parameters/covariance matrix. This would allow you to get rid of the `use_subset` argument on the sampling method since calling `sample_resonance_parameters` on the new instance would use the subset by default.

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The covariance module also has the ability to sample a new set of parameters using the covariance matrix. Currently the sampling uses np.multivariate_normal(). Because parameters are assumed to have a multivariate normal distribution this method doesn't not currently guarantee that sampled parameters will be positive."
```
np.multivariate_normal -> numpy.multivariate_normal

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,812 @@
+        items : list
+            Items from the CONT record at the start of the resonance range
+            subsection
+        file2params : openmc.data.ResonanceRange object
```
It looks like this is still actually taking a DataFrame of resonance parameters rather than the `ResonanceRange` object that owns it. I think it would be better to use the `ResonanceRange` object directly and store it as an attribute of the `ResonanceCovarianceRange` instance. That way, when `sample_resonance_parameters` is called, there would be no need to pass the parameters as they could be retrieved from the attribute.

paulromano replied:
Same comment applies to `ReichMooreCovariance`

paulromano commented on examples/jupyter/nuclear-data-resonance-covariance.ipynb:
```diff
@@ -0,0 +1,374 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The sampling routine requires the incorpotation of the `openmc.data.ResonanceRange` for the same resonance range object. This allows each sample itself to be its own `openmc.data.ResonanceRange` with a new set of parameters. Looking at some of the sampled parameters below:"
```
Typo: incorpotation

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,822 @@
+        parameters = self.parameters
+        cov = self.covariance
+        # Copy ResonanceRange object
+        res_range = copy.copy(self.file2res)
```
Right now, you're creating a single copy of the resonance range. If multiple samples are requested, this method actually returns a list with the same object repeated N times (this is evident in the Jupyter notebook where you can see that two different samples have the same parameters). You should be creating N copies of the original object and then setting the parameters appropriately for each one.

icmeyer replied:
Saw similar code being used for each formalism/mpar so thought I could pull it out front but guess copy.copy() and .append() behaved differently than I thought. Should be better now.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,822 @@
+                               'captureWidth', 'fissionWidth', 'competitiveWidth']
+                    sample_params = pd.DataFrame.from_records(records,
+                                                              columns=columns)
+                    res_range.parameters = sample_params
```
Related to the previous comment, this line overwrites the parameters on each iteration of the for loop. (also, I don't see a similar line updating the parameters for the mpar=4/5 and the RM cases)

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,743 @@
+from collections import defaultdict, MutableSequence, Iterable
```
Remove defaultdict, Iterable (not used)

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,743 @@
+    mpar : int
+        Number of parameters in covariance matrix for each individual resonance
+    formalism : str
+        String descriptor of formalism
```
Need to add description of file2res here (since it is assumed to be present for any subclass)

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,743 @@
+        self.energy_min = energy_min
+        self.energy_max = energy_max
+
+    def res_subset(self, parameter_str, bounds):
```
What do you think about just naming this `subset`? The `res` seems kind of superfluous to me.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,743 @@
+                gn = sample[1::mpar]
+                gg = sample[2::mpar]
+                gfa = sample[3::mpar] if mpar > 3 else parameters['fissionWidthA'].values
+                gfb = sample[3::mpar] if mpar > 3 else parameters['fissionWidthB'].values
```
4::mpar

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,743 @@
+
+            # Add parameters from File 2
+            parameters = _add_file2_contributions(parameters,
+                                                  resonance.parameters)
```
It looks like the last few lines here (starting with `columns = ...`) are the same for each branch of the if -- I think those could be moved out of the if altogether (so that they're not repeated). The only thing that's different is in the lcomp == 1 case, `mpar` is not set, but that looks like a bug since it is used in the call below.

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,743 @@
+
+            # Add parameters from File 2
+            parameters = _add_file2_contributions(parameters,
+                                                  resonance.parameters)
```
Same comment here -- the lines starting from `columns = ...` looks redundant on the branches of the if.

paulromano commented on tests/unit_tests/test_data_neutron.py:
```diff
@@ -39,7 +39,7 @@ def sm150():
     """Gd154 ENDF data (contains Reich Moore resonance range)"""
     filename = os.path.join(_ENDF_DATA, 'neutrons', 'n-064_Gd_154.endf')
-    return openmc.data.IncidentNeutron.from_endf(filename)
+    return openmc.data.IncidentNeutron.from_endf(filename, covariance = True)
```
`covariance=True`

paulromano commented on openmc/data/resonance_covariance.py:
```diff
@@ -0,0 +1,743 @@
+        res_cov_range.covariance = cov_subset
+        return res_cov_range
+
+    def sample_resonance_parameters(self, n_samples):
```
What do you think about naming this simply `sample()`?

icmeyer replied:
Makes sense to me. Wasn't sure how descriptive the names needed to be but as with `res_subset` I think that having them as methods of the objects provides enough context

icmeyer commented:
It looks like the build is failing with Python 3.4 on the following test. 
https://github.com/mit-crpg/openmc/blob/ac57dd929841026bb91cd1f89d67b5556b4ad157/tests/unit_tests/test_capi.py#L89-L100
Running on my machine with 3.4 gives me 
`>       assert m.densities == pytest.approx(test_dens)`
`E       ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`
Does this test need to be changed? Also odd that on some of the builds Travis isn't showing the final report of pytest
